### PR TITLE
Add support for tensorflow backend which allows for differentiability

### PR DESCRIPTION
**Pull request by chaithyagr**
I have refactored the codes and made it more meaningful.

With this PR, we can support differentiable optimization on all the algorithms.

Finally, I incorporated #111 in here with right code styles

### CODE REVIEW

**Comment by chaithyagr** [comment]
@sfarrens looks like the `develop` branch isnt in sync with `main` branch...
Can you please update it?

**Comment by sfarrens** [comment]
@chaithyagr it should be OK now. 

I had a quick look at the PR and everything looks OK to me, but I will have a closer look this week before merging.

File: modopt/base/backend.py [hunk 1]
```diff
@@ -8,35 +8,71 @@
...
+    Returns the backend module for input specified by string
+
+    Parameters
+    ----------
+    backend: str
+        String holding the backend name. One of `tensorflow`,
+        `numpy` or `cupy`.
+
+    Returns
+    -------
```
Comment (sfarrens) [inline modopt/base/backend.py:62]: You need to specify the return type

File: modopt/base/backend.py [hunk 2]
```diff
@@ -54,48 +90,47 @@ def get_array_module(input_data):
...
+    This method changes the backend of an array
+    This can be used to copy data to GPU or to CPU
 
     Parameters
     ----------
     input_data : numpy.ndarray or cupy.ndarray
         Input data array to be moved
+    backend: str, default `cupy`
+        The backend to use, one among `tensorflow`, `cupy` and
+        `numpy`
```
Comment (sfarrens) [inline modopt/base/backend.py:114]:
```suggestion
    backend: str, optional
        The backend to use, one among `tensorflow`, `cupy` and
        `numpy`. Default is `cupy`.
```

File: modopt/base/backend.py [hunk 3]
```diff
@@ -107,13 +142,20 @@ def move_to_cpu(input_data):
...
-    if xp == np:
+    if xp == LIBRARIES['numpy']:
         return input_data
-
-    return input_data.get()
+    elif xp == LIBRARIES['cupy']:
+        return input_data.get()
+    elif xp == LIBRARIES['tensorflow']:
+        return input_data.data.numpy()
+    raise ValueError('Cant identify the kind of array!')
```
Comment (sfarrens) [inline modopt/base/backend.py:158]:
```suggestion
    raise ValueError('Cannot identify the array type.')
```

File: modopt/opt/algorithms.py [hunk 4]
```diff
@@ -148,7 +132,10 @@ def metrics(self, metrics):
 
         if isinstance(metrics, type(None)):
             self._metrics = {}
-        elif not isinstance(metrics, dict):
+            return
```
Comment (sfarrens) [inline modopt/opt/algorithms.py:135]: What's the point of this `return` statement?

**Review by sfarrens — CHANGES_REQUESTED** [review]

**Comment by sfarrens** [comment]
@chaithyagr sorry I missed the comment about #111 on my first glance at this PR.

I made a few minor comments, once these are resolved I am happy to merge. I can then release a patch shortly before making the PySAP release.

File: modopt/opt/algorithms.py (hunk 4 above)
Comment (chaithyagr) [reply modopt/opt/algorithms.py:135]: Old code, removing

File: modopt/base/backend.py [hunk 5]
```diff
@@ -60,7 +60,8 @@ def get_backend(backend):
 
     Returns
     -------
-    The module for carrying out calculations
+    module
```
Comment (sfarrens) [inline modopt/base/backend.py:63]: `return LIBRARIES[backend], backend` looks like a `tuple` to me.

```suggestion
    tuple
```

**Review by sfarrens — CHANGES_REQUESTED** [review]

File: modopt/base/backend.py (hunk 5 above)
Comment (chaithyagr) [reply modopt/base/backend.py:63]: Ahhh, my bad, fixing the documentation

**Comment by sfarrens** [comment]
OK, looks like there are just a couple of minor things to fix for the tests. Once the tests pass I will merge.

**Review by sfarrens — APPROVED** [review]

**Comment by chaithyagr** [comment]
@sfarrens I have added a new ignore WPS407 in backend.py, for mutable constants. I could not code my way away from it.. Can you please let me know if there is a way that I am missing...

**Comment by sfarrens** [comment]
@chaithyagr That's fine for now. There are plenty of ignores that will need to be revisited in later issues, some of them we can justify while others we may want to rethink how we are doing things. 🙂

**Pull request merged**